src/projection.jl

```diff
@@ -310,6 +322,45 @@ function (project::ProjectTo{<:Tangent{<:Tuple}})(dx::Tuple)
     dy = map((f, x) -> f(x), project.elements, dx)
     return project_type(project)(dy...)
 end
+function (project::ProjectTo{<:Tangent{<:NamedTuple}})(dx::NamedTuple)
+    dy = _project_namedtuple(backing(project), dx)
+    return project_type(project)(; dy...)
+end
+
+# Diffractor returns not necessarily a named tuple with all keys and of the same order as
+# the projector
+# Thus we can't use `map`
+function _project_namedtuple(f::NamedTuple{fn,ft}, x::NamedTuple{xn,xt}) where {fn,ft,xn,xt}
```

Comment:
what is this function for?

Can't we just  stick the thing into a  `Tangent{typeof(f), typeof(x)}(x)` ?
which should robustly handly non-present keys and keys in different orders.
And if for some reason we can't handle that then add a `canonicalize` ?

Reply:
It is our custom projection `map`. Initially, in the first commit I justed use `map` with the named tuple of projectors and named tuple of derivatives, as suggested by @mcabbott. However, `map` requires that the names of both named tuples are exactly identical, i.e., all derivatives are present and in the same order as the projectors. This function here just maps all existing derivatives and throws a more descriptive error if a derivative is present without corresponding projector.

Reply:
I guess it could take short namedtuples and route them through Tangent -> canonise -> backing -> map -> Tangent, to re-use more stuff:
```
julia> using ChainRulesCore

julia> x = (a=1, b=2, c=3); dx = (b=400,);

julia> Tangent{typeof(x)}(; dx...)
Tangent{NamedTuple{(:a, :b, :c), Tuple{Int64, Int64, Int64}}}(b = 400,)

julia> ChainRulesCore.canonicalize(ans)
Tangent{NamedTuple{(:a, :b, :c), Tuple{Int64, Int64, Int64}}}(a = ZeroTangent(), b = 400, c = ZeroTangent())

julia> ChainRulesCore.backing(ans)
(a = ZeroTangent(), b = 400, c = ZeroTangent())
```

My slight reservation about all approaches really is whether we can insert enough complication to confuse Diffractor when it wants to take a 3rd derivative or something.

Reply:
why do we need the `backing` step?
why not
`Tangent -> map`, which already returns a `Tangent`?

Reply:
Well, like this it doesn't: 
```
julia> tang
Tangent{NamedTuple{(:a, :b, :c), Tuple{Int64, Int64, Int64}}}(a = ZeroTangent(), b = 400, c = ZeroTangent())

julia> projs = map(ProjectTo, x);

julia> map((f,x) -> f(x), projs, tang)
3-element Vector{Any}:
    ZeroTangent()
 400.0
    ZeroTangent()
```

Reply:
@devmotion what would you think of adding multi-arg map to `Tangent{T}` and using that here, rather than a more custom function?
Perhaps that wouldn't make sense since we basically would need to have the auto-inserting of ZeroTangent behavour for the non-Tangent{T} arg also?

Reply:
Hmm, I'm not sure - my feeling is that it would be good to have another use case for it. For this PR, it seems we would want to restrict it to basically just `map(f, ::NamedTuple, ::Tangent{<:NamedTuple})` and need have special needs for what should be allowed and what not.

Reply:
It would also be a bit inconsistent with `map` for `NamedTuple` defined in base which requires that the same keys are present in exactly the same order.

Reply:
This seems fine. If one of your mapped collections is a `Tangent` we should follow the `Tangent` convention for map, which doesn't require these things.
In any case, I agree we would want a second use-case, so we can do it in a follow up when and if that occurs.

Reply:
I haven't dug through all the functions closely, recently, but my reservation here is that this seems close to being a second use of `canonicalize`, just with a different carefully optimised generated implementation. It seems that if ever something breaks one, we'll have to fix both.

Is there a precedent anywhere else here about whether filling in all fields with NoTangent is preferable / not compared to leaving omitted ones out?

Reply:
We would preferably not fill in all fields.

Reply:
This was my understanding as well - and hence I don't think one should use `canonicalize` here since we don't want to fill all fields.

Reply:
@mcabbott Are you OK with merging the PR as is and improving the implementation later, if e.g. there is a clear need for a two argument `map`?

Reply:
Yes, do it. I think this is the right behaviour. I do wish it could be shorter but that's not the end of the world. Sorry about dragging this out so long.